pegasus/src/homepage.rb

```diff
@@ -0,0 +1,130 @@
+        {type: "blockshort", title: "homepage_slot_text_title_hoc", text: "homepage_slot_text_blurb_hoc", color1: "0, 173, 188", color2: "89, 202, 211", url: "/learn", image: "/images/mc/2016_homepage_hocblock.jpg"},
+        {type: "blockshort", title: "studiobar_flappy_title", text: "studiobar_flappy_body", color1: "185, 191, 21", color2: "209, 213, 103", url: CDO.studio_url("/s/flappy/reset"), image: "/shared/images/courses/logo_tall_flappy.jpg"}
+      ]
+    end
+  end
+
+  def self.get_video
+    video = get_actions.find {|a| a[:type] == "video"}
+
+    {
+      video_code: video[:youtube_id],
+      download_path: video[:download_path],
+      facebook: {u: video[:facebook]},
+      twitter: {related: 'codeorg', text: video[:twitter]}
+    }
+  end
+
+  def self.get_census(request)
+    census_announcement = DCDO.get('census_announcement', nil)
+
+    # Whether we show non-rotating background image and center options on desktop.
+    show_census_layout = request.language == "en"
+
+    # Whether we have a special announcement to also show.
+    show_census_announcement = census_announcement && census_announcement['homepage_text'] && census_announcement['homepage_subtext'] && request.language == "en"
+
+    return show_census_layout, show_census_announcement, census_announcement
+  end
+
+  def self.get_heroes_arranged(request)
+    show_census_layout, _, _ = Homepage.get_census(request)
+    census_announcement_hero = [{text: "homepage_hero_text_stat_students", centering: "50% 30%", type: "stat", textposition: "bottom", image: "/images/homepage/announcement.jpg"}]
+
+    # Generate a random set of hero images alternating between non-celeb and celeb.
+    heroes = get_heroes
+    hero_display_time = 13 * 1000
+
+    if rack_env != :production && request.params["preview"]
+      # On non-production, special "?preview=true" flag shows all heroes, and more quickly, for easier previewing
+      heroes_arranged = heroes
+      hero_display_time = 6 * 1000
+    elsif rack_env != :production && request.params["lock-hero"]
+      # For UI tests just lock to the first hero image
+      heroes_arranged = heroes[0, 1]
+    elsif show_census_layout
+      heroes_arranged = census_announcement_hero
+    else
+      # The order alternates person & stat.  Person alternates non-celeb and
+      # celeb.  Non-celeb is student or teacher. We open with a celeb, i.e.,
+      # celeb, stat, non-celeb, stat, celeb, stat, non-celeb, stat, celeb, stat,
+      # etc.
+      heroes.shuffle!
+      heroes_nonceleb = heroes.select {|hero| ["student", "teacher"].include? hero[:type]}
+      heroes_celeb = heroes.select {|hero| hero[:type] == "celeb"}
+      heroes_stat = heroes.select {|hero| hero[:type] == "stat"}
+      heroes_arranged =
+        if heroes_stat.empty?
+          heroes_nonceleb.zip(heroes_celeb).flatten.compact.first(6)
+        else
+          heroes_stat.zip(heroes_nonceleb.zip(heroes_celeb).flatten).flatten.compact.first(6)
```

Comment:
Worth extracting a `MAX_HEROES_IN_ROTATION = 6` constant at the top of the file?